pkgs/by-name/ta/tana/package.nix

```diff
@@ -0,0 +1,117 @@
+{ libX11
+, libxcb
+, libXcomposite
+, libXdamage
+, libXext
+, libXfixes
+, libXrandr
+, stdenv
+, lib
+, alsa-lib
+, at-spi2-atk
+, atkmm
+, cairo
+, cups
+, dbus
+, expat
+, glib
+, gtk3
+, libdrm
+, libglvnd
+, libxkbcommon
+, mesa
+, nspr
+, nss
+, pango
+, systemd
+, fetchurl
+, autoPatchelfHook
+, dpkg
+, makeWrapper
+}:
+let
+  glLibs = [ libglvnd mesa ];
+  libs = [
+    alsa-lib
+    atkmm
+    at-spi2-atk
+    cairo
+    cups
+    dbus
+    expat
+    glib
+    gtk3
+    libdrm
+    libX11
+    libxcb
+    libXcomposite
+    libXdamage
+    libXext
+    libXfixes
+    libxkbcommon
+    libXrandr
+    nspr
+    nss
+    pango
+  ];
+  buildInputs = glLibs ++ libs;
+  version = "1.0.15";
+  pname = "tana";
+in
+stdenv.mkDerivation {
+  inherit pname version buildInputs;
+
+  src = fetchurl {
+    url = "https://github.com/tanainc/tana-desktop-releases/releases/download/v${version}/tana_${version}_amd64.deb";
+    hash = "sha256-94AyAwNFN5FCol97US1Pv8IN1+WMRA3St9kL2w+9FJU=";
+  };
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    dpkg
+    makeWrapper
+  ];
+
+  # Needed for Zygote
+  runtimeDependencies = [
+    systemd
+  ];
+
+  dontConfigure = true;
+  dontBuild = true;
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out
+    cp -r usr/* $out
+    rm $out/bin/tana
+    runHook postInstall
+  '';
+
+  postFixup = ''
+    substituteInPlace $out/share/applications/tana.desktop \
+      --replace "Exec=tana" "Exec=$out/bin/tana" \
+      --replace "Name=tana" "Name=Tana"
+    makeWrapper $out/lib/tana/Tana $out/bin/tana \
+      --set LD_LIBRARY_PATH ${lib.makeLibraryPath glLibs} \
+      --suffix LD_LIBRARY_PATH : $out/lib/tana
```

Comment:
this is not needed when using autopatchelfhook.
to get autopatchelf hook to patch the .so files with an rpath use:

```nix
# add RPATHs to all shared objects
appendRunpaths = map (pkg: (lib.getLib pkg) + "/lib") [
];
# add RPATH to executable only
runtimeDependencies = [
];
```

see: https://github.com/eliandoran/nixpkgs/blob/e63e189658a3c0c829a8403c4b54e4f2e2fcd857/pkgs/by-name/lu/lunacy/package.nix#L52-L70 for an example.

Reply:
The more you do, the more you learn! Thanks for pointing this out, didn't know that!

It should be fixed now (I could not remove the first `LD_LIBRARY_PATH` as the GL stuff is `dlopen()` at runtime, and removing the wrapper completely made some weird GL errors)

Reply:
if the RPATHS are correct dlopen will work fine. having to use `LD_LIBRARY_PATH` means that the RPATHs are not correct so something was missing.

if you use:
```nix
# add RPATHs to all shared objects
appendRunpaths = map (pkg: "${lib.getLib pkg}/lib") glLibs ++ [ "${placeholder "out"}/lib/tana" ];
```
that should be equivalent to what you have in `LD_LIBRARY_PATH`

Reply:
though i guess if there are libraries doing a dlopen that are not in the derivation they could be picking up things in the `LD_LIBRARY_PATH`

Reply:
> if the RPATHS are correct dlopen will work fine. having to use `LD_LIBRARY_PATH` means that the RPATHs are not correct so something was missing.
> 
> if you use:
> 
> ```nix
> # add RPATHs to all shared objects
> appendRunpaths = map (pkg: "${lib.getLib pkg}/lib") glLibs ++ [ "${placeholder "out"}/lib/tana" ];
> ```
> 
> that should be equivalent to what you have in `LD_LIBRARY_PATH`

It is indeed working (I modified the derivation once again). I don't know what I was doing wrong before. Even though I have an idea of what's happening behind the scenes, the amount of black magic involved in the autoPatchelfHook is amazing.

Thanks a lot for your inputs! I removed the wrapper completely and reverted back to the default symbolic link which was present in the original package.

Reply:
yeah, appendRunpaths is not documented and is needed when libraries are performing dlopens as autopatchelfhook can not detect that from just reading the elf metadata.

Reply:
using `LD_LIBRARY_PATH` isn't the worst, but it has a global effect for all shared objects and all processes, which is not desirable.

for example -- when building libxml2.2 they set `LD_LIBRARY_PATH` to point to the new `libxm2.2` for tests but also run the compiler, which is linked to libxml2.2 but picks up the just built libxml2.2 and then just breaks.

so it is good to avoid it if possible.